Transition to API version 0.0.4

spell00 opened the pull request:
## Brief summary of changes
Transition to v0.0.4-dev to include new extensions of the API that are not in the current specs.
The changes include:
- The specs of v0.0.3 is copied into v0.0.4-dev
- The integration tests are changed to test v0.0.4-dev
- All current endpoints allow v0.0.4-dev of the API

#### Link(s) to related issue(s)

* Resolves #6941

spell00 commented:
The version new name needs to be discussed.

spell00 commented:
@xlecours Could you please have a look at this?

xlecours reviewed (changes requested):
Could you add something in CHANGELOG.md to mention that there is a new version of the api specification in the modules section. We will add items under it for each new or changed endpoints.

xlecours commented:
@driusan This PR set the table for additional endpoints and modifications of existing endpoints. I will evaluate if the following PR can be include in v0.0.4
https://github.com/aces/Loris/pull/6905
https://github.com/aces/Loris/pull/6899
https://github.com/aces/Loris/pull/6780
https://github.com/aces/Loris/pull/6775
https://github.com/aces/Loris/pull/6161
https://github.com/aces/Loris/pull/5354

Meanwhile, I think this PR is the first step. Would you prefer to have a feature branch instead?

spell00 commented on CHANGELOG.md:
```diff
@@ -120,4 +122,4 @@ be used by projects having custom modules not in LORIS. (PR #5913)
 - Dashboard was refactored to turn panels into module widgets. (PR #5896)
-- Add CSSGrid component type (PR #6090)
\ No newline at end of file
+- Add CSSGrid component type (PR #6090)
```
I don't see why this is highlighted as a change, it looks like both lines are exactly the same

xlecours replied:
I don't know what changed but copy-pasting the line seems to propose a change.

spell00 replied:
it's impossible to commit the suggestion though, it says `this suggestion is invalid because no changes were made to the code`

xlecours commented on CHANGELOG.md:
```diff
@@ -120,4 +122,4 @@ be used by projects having custom modules not in LORIS. (PR #5913)
 - Dashboard was refactored to turn panels into module widgets. (PR #5896)
-- Add CSSGrid component type (PR #6090)
\ No newline at end of file
+- Add CSSGrid component type (PR #6090)
```
```suggestion
- Add CSSGrid component type (PR #6090)
```

driusan reviewed (changes requested):
Since there's still other PRs to go into 0.0.4, please add the -dev suffix to the version number until it's ready. Otherwise we'll have multiple incompatible versions of "0.0.4".

Other than that it looks good to me.

xlecours commented:
> Since there's still other PRs to go into 0.0.4, please add the -dev suffix to the version number until it's ready. Otherwise we'll have multiple incompatible versions of "0.0.4".
> 
> Other than that it looks good to me.

@driusan it is not released.It is implicitly `-dev`. Otherwise, we would need to change all the `0.0.4` to `0.0.4-dev` then back to `0.0.4` before the release.

spell00 commented:
@xlecours @driusan  So what should it be? Is `-dev` because there is not a new version of the API at each release? How is it decided when it is time to remove the `-dev` suffix and start a new version? It's an easy and fast update to do, I wonder if updating the version systematically on each release wouldn't be faster and more straightforward, as it would avoid having a discussion on every release about removing or adding `-dev`? 

Also, I thought it was implicitly `-dev` too, because it is `v0`. Wouldn't it make sense if all official new version of the API would update the first number (e.g. v0.0.4 -> v1.0.0) and all updates of the last number would be implicitely a `dev` version (e.g. v1.0.0  -> 1.0.1 -> 1.0.2)?

christinerogers commented:
@driusan i think the naming is now resolved per today's loris meeting. 
Could you please re-review? 
 this PR is holding up several other ones and will be key for release.

spell00 commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
```diff
@@ -1,5 +1,4 @@
-# Loris API - v0.0.2
```
It looks like I modified a lot of things in the doc of v0.0.2, but it is actually modified from v0.0.3. I removed `LorisRESTAPI.md` and added `LorisRESTAPI_v0.0.4-dev.md`. 

I did not modifiy the original `LorisRESTAPI_v0.0.3.md` so Github must consider I replaced `v0.0.2` by `v0.0.4-dev`.

spell00 replied:
I solved this problem by renaming `LorisRESTAPI_v0.0.3.md` to `LorisRESTAPI.md`.

spell00 commented on modules/api/php/endpoints/candidate/visit/visit.class.inc:
```diff
@@ -75,8 +75,10 @@ class Visit extends Endpoint implements \LORIS\Middleware\ETagCalculator
      */
     protected function supportedVersions() : array
     {
-        // Removed 0.0.2 since session requires a project.
```
this comment is no longer relevant, since 0.0.2 is no longer supported anywhere

spell00 commented:
@driusan @xlecours as per our discussion this morning, I have removed support for `v0.0.2` in all endpoints. I also removed the doc for `v0.0.2`.

I am not sure if @xlecours agreed with the use of the `-dev` suffixe, but I think we should proceed with this PR as it is (with `-dev`). If we were to decide to change the version to remove `-dev`, it is very fast to change it.

From the directory `modules/api/php/endpoints/`, I would just execute the command line:

`grep -rli 'v0.0.4-dev' * | xargs -i@ sed -i 's/v0.0.4-dev/v0.0.4/g' @`

cc @christinerogers

christinerogers commented:
@xlecours @driusan could we get some joint attention on this PR? It's blocking the rest of Simon's work from moving forward, and this needs to be resolved in time.

xlecours commented:
ok @driusan let's move forward.
